File: regions/system/library/swift/pagination/Tests/K1s0PaginationTests/PaginationTests.swift
import Testing
@testable import K1s0Pagination

@Suite("Pagination Tests")
struct PaginationTests {
    @Test("PageRequestが正しく初期化されること")
    func testPageRequest() {
        let req = PageRequest(page: 1, perPage: 10)
        #expect(req.page == 1)
        #expect(req.perPage == 10)
    }

    @Test("PageResponseが正しく計算されること")
    func testPageResponseCreate() {
        let req = PageRequest(page: 1, perPage: 10)
        let resp = PageResponse<String>.create(items: ["a", "b"], total: 25, request: req)
        #expect(resp.items.count == 2)
        #expect(resp.total == 25)
        #expect(resp.totalPages == 3)
    }

    @Test("perPageが0の場合totalPagesが0になること")
    func testZeroPerPage() {
        let req = PageRequest(page: 1, perPage: 0)
        let resp = PageResponse<String>.create(items: [], total: 10, request: req)
        #expect(resp.totalPages == 0)
    }

    @Test("totalが0の場合totalPagesが0になること")
    func testZeroTotal() {
        let req = PageRequest(page: 1, perPage: 10)
        let resp = PageResponse<String>.create(items: [], total: 0, request: req)
        #expect(resp.totalPages == 0)
    }

    @Test("カーソルのエンコードとデコードが正しく動作すること")
    func testCursorRoundTrip() throws {
        let sortKey = "2024-01-15"
        let id = "item-123"
        let encoded = encodeCursor(sortKey: sortKey, id: id)
        let (decodedSortKey, decodedId) = try decodeCursor(encoded)
        #expect(decodedSortKey == sortKey)
        #expect(decodedId == id)
    }

    @Test("無効なカーソルがエラーになること")
    func testInvalidCursor() {
        #expect(throws: CursorError.self) {
            try decodeCursor("!!!invalid-base64!!!")
        }
    }

    @Test("CursorRequestフィールド")
    func testCursorRequest() {
        let req = CursorRequest(cursor: "abc", limit: 20)
        #expect(req.cursor == "abc")
        #expect(req.limit == 20)
    }

    @Test("CursorMetaフィールド")
    func testCursorMeta() {
        let meta = CursorMeta(nextCursor: "next", hasMore: true)
        #expect(meta.nextCursor == "next")
        #expect(meta.hasMore == true)
    }

    @Test("PaginationMetaフィールド")
    func testPaginationMeta() {
        let meta = PaginationMeta(total: 100, page: 2, perPage: 10, totalPages: 10)
        #expect(meta.total == 100)
        #expect(meta.page == 2)
        #expect(meta.perPage == 10)
        #expect(meta.totalPages == 10)
    }

    @Test("validatePerPage有効値")
    func testValidatePerPageValid() throws {
        #expect(try validatePerPage(1) == 1)
        #expect(try validatePerPage(50) == 50)
        #expect(try validatePerPage(100) == 100)
    }

    @Test("validatePerPageゼロ")
    func testValidatePerPageZero() {
        #expect(throws: PerPageValidationError.self) {
            try validatePerPage(0)
        }
    }

    @Test("validatePerPage最大超過")
    func testValidatePerPageOverMax() {
        #expect(throws: PerPageValidationError.self) {
            try validatePerPage(101)
        }
    }
}

File: regions/system/library/csharp/pagination/tests/PaginationTests.cs
using K1s0.System.Pagination;

namespace K1s0.System.Pagination.Tests;

public class PaginationTests
{
    [Fact]
    public void PageResponse_Create_CalculatesTotalPages()
    {
        var items = new List<string> { "a", "b" };
        var req = new PageRequest(1, 2);
        var resp = PageResponse<string>.Create(items, 5, req);

        Assert.Equal(5, resp.Total);
        Assert.Equal(1u, resp.Page);
        Assert.Equal(2u, resp.PerPage);
        Assert.Equal(3u, resp.TotalPages);
        Assert.Equal(2, resp.Items.Count);
    }

    [Fact]
    public void PageResponse_Create_ExactDivision()
    {
        var items = new List<int> { 1, 2 };
        var req = new PageRequest(1, 2);
        var resp = PageResponse<int>.Create(items, 4, req);

        Assert.Equal(2u, resp.TotalPages);
    }

    [Fact]
    public void PageResponse_Create_ZeroPerPage_ReturnsZeroTotalPages()
    {
        var items = new List<int>();
        var req = new PageRequest(1, 0);
        var resp = PageResponse<int>.Create(items, 10, req);

        Assert.Equal(0u, resp.TotalPages);
    }

    [Fact]
    public void CursorPagination_EncodeAndDecode_Roundtrip()
    {
        var sortKey = "2024-01-15";
        var id = "item-12345";
        var cursor = CursorPagination.Encode(sortKey, id);
        var (decodedSortKey, decodedId) = CursorPagination.Decode(cursor);

        Assert.Equal(sortKey, decodedSortKey);
        Assert.Equal(id, decodedId);
    }

    [Fact]
    public void CursorPagination_Decode_MissingSeparator_Throws()
    {
        var cursor = Convert.ToBase64String(global::System.Text.Encoding.UTF8.GetBytes("noseparator"));
        Assert.Throws<FormatException>(() => CursorPagination.Decode(cursor));
    }

    [Fact]
    public void PageRequest_RecordEquality()
    {
        var a = new PageRequest(1, 10);
        var b = new PageRequest(1, 10);
        Assert.Equal(a, b);
    }

    [Fact]
    public void CursorRequest_Fields()
    {
        var req = new CursorRequest("abc", 20);
        Assert.Equal("abc", req.Cursor);
        Assert.Equal(20u, req.Limit);
    }

    [Fact]
    public void CursorMeta_Fields()
    {
        var meta = new CursorMeta("next", true);
        Assert.Equal("next", meta.NextCursor);
        Assert.True(meta.HasMore);
    }

    [Fact]
    public void PaginationMeta_Fields()
    {
        var meta = new PaginationMeta(100, 2, 10, 10);
        Assert.Equal(100, meta.Total);
        Assert.Equal(2u, meta.Page);
        Assert.Equal(10u, meta.PerPage);
        Assert.Equal(10u, meta.TotalPages);
    }

    [Fact]
    public void PerPageValidator_ValidValues()
    {
        Assert.Equal(1u, PerPageValidator.Validate(1));
        Assert.Equal(50u, PerPageValidator.Validate(50));
        Assert.Equal(100u, PerPageValidator.Validate(100));
    }

    [Fact]
    public void PerPageValidator_Zero_Throws()
    {
        Assert.Throws<ArgumentOutOfRangeException>(() => PerPageValidator.Validate(0));
    }

    [Fact]
    public void PerPageValidator_OverMax_Throws()
    {
        Assert.Throws<ArgumentOutOfRangeException>(() => PerPageValidator.Validate(101));
    }
}
